### ISSUE TITLE
Disallow ambiguous overloads by default and allow controlled overload refinement

### ISSUE THREAD

**Issue by jckarter** [jckarter/clay#383]
As suggested by @stepancheg, overloads should not be universally overrideable, because overlapping overloads are often a mistake and can have unintended consequence, and there is not a well-defined ordering between overloads.

**Comment by ghost** [comment]
I've had a go at implementing final overloads, however there a couple of issues I'd like bring up. 

Firstly, what do we need to happen for overloads using pattern names e.g.

```
[T when Float?(T) and Integer?(U)]
overload T(x:U) = . . . blah . . . 
```

when including the overload pattern list I get recursion. I haven't found a solution to this yet. Should pattern overloads be final?

Also, final overloads have led to the following 'override' pattern (heavily used throughout the lib, so checking before spending hours modifying everything):

```
[T] SignedInteger?(#T) = false; // Implicit overload - this is final and matches everything
override SignedInteger?(#Int8) : Bool = true; // so everything else has to be an 'override'
override SignedInteger?(#Int16) : Bool = true;
override SignedInteger?(#Int32) : Bool = true;
override SignedInteger?(#Int64) : Bool = true;
override SignedInteger?(#Int128) : Bool = true;

```

**Comment by jckarter** [comment]
I don't understand your first problem. How is there recursion where there wasn't originally? How does your final overload mechanism work?

In the second case, I think you have the keyword in the wrong place. It'd be better to keyword the overridable case than the overriding case. Maybe `default`? For example:

```
[T] define SignedInteger?(#T) : Bool;
[T] default overload SignedInteger?(#T) = false;
overload SignedInteger?(#Int8) = true;
...
```

**Comment by jckarter** [comment]
Another thing is, for patterns at least, there is a somewhat natural linearization order:

```
[T, U] overload foo(x:T, y:U);
// is less specific than
[T, U when Something?()] overload foo(x:T, x:U);
// is less specific than
[T, U] overload foo(x:T, y:Foo[U]);
// is less specific than
[T] overload foo(x:T, y:Foo[Bar]);
// is less specific than
overload foo(x:Bas, y:Foo[Bar]);
// is less specific than
overload foo(ref x:Bas, rvalue y:Foo[Bar]);
```

Overloads could be sorted by pattern specificity by default and do what you want probably 99% of the time.

**Comment by ghost** [comment]
I've updated my implementation to follow stepanchegs suggestion rather than make final the default.
The recursion issue I mentioned was due to restarting the conflicting overload check of the same function during predicate evaluation of constructor overloads. Easily fixed. 

My implementation is quite simple.
1. lookup entry in the overload list (match index is already stored)
2. if entry not final check for conflicting final overload in the rest of the list
3. if second entry found return an error

e.g.

```
define foo;
overload foo(x) = false;
overload foo(y) = true; // this is OK
```

and

```
define foo;
overload foo(x) = false;
final overload foo(z) = true // this is OK
```

and

```
define foo;
final overload foo(x) = false;
overload foo(z) = true // this returns final overload error
```

**Comment by jckarter** [comment]
Sounds good. To be clear—"conflicting overload" can be any overload that matches the same input types, not just an overload with an identical signature, right? So something like this would raise an error:

```
define foo;
final overload foo(x) = 0;
final overload foo(x:Int) = 1;

main() {
    foo('1'); // ok, only foo(x) matches Char
    foo(1); // error, both foo(x) and foo(x:Int) match Int
}
```

In the absence of a smarter base linearization, I can see that having non-`final` be default is more convenient. Compile-time performance would probably also be adversely affected by having to check all overloads for overlap at every call site.

**Comment by ghost** [comment]
I did try checking all overloads for every call, the performance hit was huge, ~94 seconds to compile hello world example.

Yes, i've used a full matchInvoke on final overload checks, so your example would raise an error.

**Comment by ghost** [comment]
Non-final by default seems like a poor solution, I think it may be better to try and improve the overload resolution mechanism and use an explicit `default/open` case as @jckarter  suggested. Maybe a `final/close` attribute could be used to close off an open overload e.g.

```
[T] define SignedInteger?(#T) : Bool;
[T] open overload SignedInteger?(#T) = false;
overload SignedInteger?(#Int8) = true;
overload SignedInteger?(#Int16) = true;
close overload SignedInteger?(#Int32) = true;
```

Not so sure that's a sensible suggestion, but it seems like it may be a useful pattern for generics.

**Comment by stepancheg** [comment]
I like `open/close` modifiers, and I like final to be default.

I have two improvement proposals.

#### 1

Probably `close` should be a separate statement, not an overload function modifier.

```
[T] define SignedInteger?(#T) : Bool;
[T] open overload SignedInteger?(#T) = false;
overload SignedInteger?(#Int8) = true;
overload SignedInteger?(#Int16) = true;
overload SignedInteger?(#Int32) = true;
close overload SignedInteger?;
```

Because `close` prohibits overloading all functions, not just the one which has the modifier. I think that separate `close` statement would make the feature easier to understand.

#### 2

And I also suggest, that only `defines` should be overloadable. E. g. this code should be invalid:

```
// assuming final is default
foo() = 1;
overload foo(a) = 2;
```

(this code is valid now and valid after previous `open/close` proposals, because `foo` overload does not conflict with first `foo` declaration)

**Comment by jckarter** [comment]
I agree with #2; it would be nice to require overloadable generics to be explicitly defined as such. #1 doesn't make sense to me; it needs two keywords and a level of micromanagement you shouldn't need to employ normally.

**Comment by ghost** [comment]
I agree with @stepancheg's 2nd proposal, maybe this should be opened as a separate issue.

**Comment by ghost** [comment]
There seems to be no clear consensus on how to handle overload ambiguity. Do we want final by default with the option  to declare an "open" case?  If so, is it wise to leave functions open, say if they are defined for a closed set of types in a library?

**Comment by stepancheg** [comment]
I think final should be default. But probably for simplicity of implementation, open default and explicit final should be developed first (to not rewrite the library right now), and open and final exchanged later.

**Comment by jckarter** [comment]
Opened #406 for making normal functions non-overloadable. I agree with stepancheg that, ideally, `final` ought to be default; overlapping overloads are usually a bug, and when you do want to overlap an existing overload, there's usually a general-specific relationship implied. (In more complex cases of overload overlap, it's good practice to resolve overlapping conditions explicitly in the predicates anyway.) I think you could get a lot of mileage with a "not-final" keyword to specify a default case, followed by specific cases:

```
default overload double(x) = x + x;
[I when Integer?(I)] overload double(x:I) = x << 1;
```

As @agemogolk said before, having normal overloads all be final greatly increases compile time in order to check for validity, so I don't know if that's practical.

**Comment by ghost** [comment]
I'm no longer working on an implementation of "open/final/default/whatever". Anybody else feel free to take ownership.

**Comment by ghost** [comment]
Wasn't it that even the overload of final definitions should work (monkeypatching)? at least I remember so from the chatlogs. thus, isn't it better if overloading from a different module is controled by a compiler flag i.e --mpatch=off/on?

i am rather confused by the above expressed ideas. for me there is a strict difference between monkeypatching (overwriting) an existing overload and extending a definition with further overloads. possibly, there should not be extra close/open keywords but only the two main keywords 'extend' for normal adding of overloads and 'overload' or 'overwrite' for _patching_/shadowing an existing overload/signature. useless to say that I dislike the term 'overload'.

**Comment by ghost** [comment]
Resolved in #478.
